samples/test/models.test.js

```diff
@@ -0,0 +1,150 @@
+          IFNULL(device.operatingSystem, "") AS os,
+          device.isMobile AS is_mobile,
+          IFNULL(geoNetwork.country, "") AS country,
+          IFNULL(totals.pageviews, 0) AS pageviews
+        FROM
+          \`bigquery-public-data.google_analytics_sample.ga_sessions_*\`
+        WHERE
+          _TABLE_SUFFIX BETWEEN '20160801' AND '20170631'
+        LIMIT  100000;`;
+
+    const datasetOptions = {
+      location: 'US',
+    };
+
+    const queryOptions = {
+      query: query,
+    };
+
+    await bigquery.createDataset(datasetId, datasetOptions);
+
+    // Run query to create a model
+    const [job] = await bigquery.createQueryJob(queryOptions);
+
+    // Wait for the query to finish
+    await job.getQueryResults();
+  });
+
+  after(async () => {
+    await bigquery
+      .dataset(datasetId)
+      .delete({force: true})
+      .catch(console.warn);
+  });
+
+  it(`should retrieve a model if it exists`, async () => {
+    const output = execSync(`node getModel.js ${datasetId} ${modelId}`);
+    assert.match(output, /Model:/);
+    assert.match(output, new RegExp(datasetId));
+    assert.match(output, new RegExp(modelId));
+  });
+
+  it(`should list models`, async () => {
+    const output = execSync(`node listModels.js ${datasetId}`);
+    assert.match(output, /Models:/);
+    assert.match(output, new RegExp(datasetId));
+  });
+
+  it(`should list models streaming`, async () => {
+    const output = execSync(`node getModel.js ${datasetId} ${modelId}`);
+    assert.match(output, new RegExp(modelId));
+  });
+
+  it(`should update model's metadata`, async () => {
+    const output = execSync(`node updateModel.js ${datasetId} ${modelId}`);
+    assert.match(
+      output,
+      new RegExp(`${modelId} description: A really great model.`)
+    );
+  });
+});
+
+describe(`Delete Model`, () => {
+  const datasetId = `gcloud_tests_${uuid.v4()}`.replace(/-/gi, '_');
+  const modelId = `gcloud_tests_${uuid.v4()}`.replace(/-/gi, '_');
+
+  before(async () => {
+    const query = `CREATE OR REPLACE MODEL \`${datasetId}.${modelId}\` 
+         OPTIONS(model_type='logistic_reg') AS
+         SELECT
+           IF(totals.transactions IS NULL, 0, 1) AS label,
+           IFNULL(device.operatingSystem, "") AS os,
+           device.isMobile AS is_mobile,
+           IFNULL(geoNetwork.country, "") AS country,
+           IFNULL(totals.pageviews, 0) AS pageviews
+         FROM
+           \`bigquery-public-data.google_analytics_sample.ga_sessions_*\`
+         WHERE
+           _TABLE_SUFFIX BETWEEN '20160801' AND '20170631'
+         LIMIT  100000;`;
+
+    const datasetOptions = {
+      location: 'US',
+    };
+
+    const queryOptions = {
+      query: query,
+    };
+
+    await bigquery.createDataset(datasetId, datasetOptions);
+
+    // Run query to create a model
+    const [job] = await bigquery.createQueryJob(queryOptions);
+
+    // Wait for the query to finish
+    await job.getQueryResults();
+  });
+
+  after(async () => {
```

Comment:
If the delete model test fails, will the model be cleaned up? Does that matter?

Reply:
Hooks run regardless of whether or not a test passes or fails. Since models are tied to datasets, this hooks should clean up any models created in this spec.